src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -1171,6 +1200,37 @@ public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object
       throw failures.failure(info, elementsShouldSatisfyExactly(actual, unsatisfiedRequirements, info));
 
   }
+  
+  /**
+   * Verifies that the actual iterable has the same size as the given consumers and 
+   * that at least one permutation of elements in the iterable exists that satisfies 
+   * the individual consumers in order.
+   *
+   * @param info contains information about the assertion.
+   * @param actual the given {@code Iterable}.
+   * @param consumers the consumers that are expected to be satisfied by the elements of the given {@code Iterable}.
+   * @throws NullPointerException if the given consumers array or any consumer is {@code null}.
+   * @throws AssertionError if the given {@code Iterable} is {@code null}.
+   * @throws AssertionError if the given {@code Iterable} and consumers don't have the same size.
+   * @throws AssertionError if any {@code Consumer} in the consumers cannot be satisfied by elements in the given {@code Iterable}.
+   */
+  public <E> void assertSatisfiesExactlyInAnyOrder(AssertionInfo info, Iterable<? extends E> actual, @SuppressWarnings("unchecked") Consumer<? super E>... consumers) {
+    assertNotNull(info, actual);
+    requireNonNull(consumers, "The Consumer<? super E>... expressing the assertions consumers must not be null");
+    for (Consumer<? super E> consumer : consumers)
+      requireNonNull(consumer, "The element in consumers must not be null");
+
+    checkSizes(actual, sizeOf(actual), consumers.length, info);
+
+    List<E>[] satisfiedElementsLists = stream(consumers)
+                                                        .map(listFilteredBySatisfiedElements(actual))
+                                                        .<List<E>> toArray(List[]::new);
```

Comment:
the code is a bit complex (the issue is too to be fair), I feel it could simplified by introducing a simple class to capture the elements satisfying a consumer, something like:
```java
// boilerplate code omitted
static class ElementsSatisfyingRequirements {
  List<E> elements;
  Consumer<? super E> requirements; 
}
```
it will make the code more intention explicit.

I would also suggest  `listFilteredBySatisfiedElements` to return directly a `List<ElementsSatisfyingRequirements>`

Reply:
Thanks, that sounds like a good idea. I will try to refactor the code accordingly.

Reply:
@joel-costigliola, would it be okay to place the ```ElementsSatisfyingRequirements``` class in a separate public class (instead of a nested class of ```Iterables```, as your previous comment indicates)?

Reply:
A separate class is fine, try to keep it package private if possible

Reply:
@joel-costigliola, I have refactored it and hope the intention is now clearer. Based on other methods and variable names, I have named to class ```ElementsSatisfyingConsumer```. Please have a look at it and let me know if you still prefer ```ElementsSatisfyingRequirements```.